irohad/consensus/yac/impl/cluster_order.cpp

```diff
@@ -10,17 +10,35 @@ namespace iroha {
     namespace yac {
 
       boost::optional<ClusterOrdering> ClusterOrdering::create(
-          const std::vector<std::shared_ptr<shared_model::interface::Peer>>
-              &order) {
+          const std::vector<std::shared_ptr<shared_model::interface::Peer>> &order,
+          std::vector<size_t> const& peer_positions) {
+        if (order.empty()) {
+          return boost::none;
+        }
+        return ClusterOrdering(order, peer_positions);
+      }
+
+      boost::optional<ClusterOrdering> ClusterOrdering::create(
+          const std::vector<std::shared_ptr<shared_model::interface::Peer>> &order) {
         if (order.empty()) {
           return boost::none;
         }
         return ClusterOrdering(order);
       }
 
       ClusterOrdering::ClusterOrdering(
-          std::vector<std::shared_ptr<shared_model::interface::Peer>> order)
-          : order_(std::move(order)) {}
+          std::vector<std::shared_ptr<shared_model::interface::Peer>> const& order,
+          std::vector<size_t> const& peer_positions) {
+
+        order_.reserve(order.size());
+        for (auto const& i : peer_positions) {
+          order_.emplace_back(order[i]);
+        }
```

Comment:
why not do this at the caller site? this way we would avoid new constructors.

Reply:
The caller must provide data for the callee as quickly and cheaply as possible. The callee must decide what to do with the data, because only it knows what it wants to do with these data(copy, modify or do nothing). Transferring logic to a higher level will lead to a change in the area of responsibility and potential overheads in performance when designing classes, when the receiving side does not want to do anything, but a copy of the data has been created for it. 

The constructors were called so, it was just done a few levels higher.
// auto order = orderer_->getOrdering(current_hash_, event.ledger_state->ledger_peers);